### DISCUSSION TITLE
Best way to convert an Image buffer `Uint8List` to `cv.Mat`

### DISCUSSION THREAD

**Discussion by gau-nernst** [rainyl/opencv_dart#18] [Q&A]
Typically we can access raw image data from other library as an `Uint8List` object. Assuming we know the format of the image (e.g. RGB uint8), what is the best (fast and convenient) way to convert it to OpenCV `Mat`? Some of the approaches I am thinking of:

- Create an empty `Mat` with the required size. Iterate over `Uint8List` and copy data over. For some formats like YUV_420sp, it might be a bit tricky to copy the data correctly.
- Encode in a lossless format (e.g. PNG) and use `cv.imdecode()` -> seems to be the most convenient, but there is overhead in compression-decompression.
- Allocate C memory with Dart FFI, copy data over, create `Mat` from C pointer. Need to release the memory after use.

**Comment by rainyl** [comment]
## UPDATE 2026.3.7

Since v2.0, the recommended ways are `Mat.fromList`, `Mat.fromBuffer`, `Mat.fromVec`, refer to the constructors of `Mat` class at https://pub.dev/documentation/dartcv4/latest/dartcv.core/Mat-class.html#constructors

## UPDATE 2024.8.30:

Since v1.2.0, the recommended ways are [Mat.fromVec](https://pub.dev/documentation/opencv_dart/latest/cv.core/Mat/Mat.fromVec.html) and [Mat.fromList](https://pub.dev/documentation/opencv_dart/latest/cv.core/Mat/Mat.fromList.html) 

## Original:

If I'm not mistaken, the `raw image data` contains image header infos, just like the data read from file directly, if so, I'm afraid the only way is  `cv.imdecode()`, Mat object in c++ opencv need to be constructed by constructors or methods like cv::imread, which is different in python (on python level, all Mat object are actually numpy.ndarray, so can be passed to opencv functions directly, but even so, it most likely a new Mat object will be created in native codes, I didn't dive into it, so just "most likely").

But if the `raw image data` are raw pixel values, it is possible to construct from values, for now the construction from UInt8List is not implemented, but the following procedure is ok:
```dart
// 2D  or 3D raw pix values
final data = ...;
final int rows = ...;
final int cols = ...;
final m = cv.Mat.create(rows, cols, type);
for (int i = 0; i < rows; i++){
  for (int j = 0; j < rows; j++){
    m.setValue<int>(i, j, data[i][j]);
  }
}
```

So, I think the best way is cv.imdecode(), please point it out if I'm wrong.

**Reply by gau-nernst** [reply]
I mean "raw image data" as raw pixel values, sorry for the confusion. Your suggested procedure is what I have in mind as the first approach that I mentioned above. Was wondering if there is a way to pass image buffer directly to OpenCV without copy.

**Reply by rainyl** [reply]
Well, the UInt8List is a dart object, so at least one copy required, I noticed the constructor 
`Mat (int rows, int cols, int type, void *data, size_t step=AUTO_STEP)`
https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#a51615ebf17a64c968df0bf49b4de6a3a

So, it seems possible to construct from data directly, I will try to implement it in the future.

**Reply by rainyl** [reply]
I am going to transfer this disscussion to issue to track it, hope you won't mind it.

**Reply by gau-nernst** [reply]
I think the main problem is how to pass "ownership" of Uint8List to OpenCV, make sure memory can be garbage-collected, and possibly don't allow anyone to mutate the buffer once it is passed to OpenCV? It is probably not trivial. Let me know if I can also help in any way.

**Reply by rainyl** [reply]
Sure, thanks~

**Reply by rainyl** [reply]
> I think the main problem is how to pass "ownership" of Uint8List to OpenCV, make sure memory can be garbage-collected

Did you mean that pass a dart UInt8List object to C/C++ directly without copying data? If so, I think it is neither easy nor necessary, if pass a UInt8List to c++, the list needs to be immutable, dart provides ImmutableList but it seems copying data is also unavoidable, so why not just copy to a native uint8 array, and unless your business really/extremely sensitive to performance, it is not necessary nowadays...

I have implemented mat construction from a list (`Mat.fromList()`), with copying to a native array, but I am sorry that I have no more time to do the above works, if you really need it, you can work on it and contributions are welcome!

**Reply by gau-nernst** [reply]
No problem at all. Indeed copy is unavoidable. Right now I have best experience with the following code

```dart
cv.Mat imageCv = cv.Mat.create(...)  // set correct width, height, and number of channels
imageCv.data.setAll(0, buffer)  // where buffer is Uint8List
```

**Reply by rainyl** [reply]
Okay.

BTW, I am confused why you set data with buffer, as far as I know, `data` is a getter of Mat and will get data from native pointer directly, so change the elements inside it won't change the data of mat.

**Reply by gau-nernst** [reply]
It works for me. You can try on your own to see. `cv.Mat.data` is `Uint8List` in Dart, which allows me to use `.setAll()` (or `.setRange()`) to copy the data to OpenCv Mat directly. This is significantly faster than using `cv.Mat.setValue()`, probably because `.setValue()` is an FFI call which has a lot of overhead.

**Reply by rainyl** [reply]
okay. I read the doc of astypedlist and noticed that it will return a reference, so it makes sense.😂

**Reply by TheoNanu** [reply]
Hello, I am struggling to apply a gaussian blur on an image from the camera. Can you please provide a full example on how to get from the image as a CameraImage object to a opencv Mat object and convert it to BGR format afterwards? This is the code I am using right now and it seems that the number of channels for converting from YUV to BGR is wrong(the image format passed to the camera controller is yuv420).

`

    Result processImage(CameraImage image) {

        final Uint8List yPlane = image.planes[0].bytes;
        final Uint8List uPlane = image.planes[1].bytes;
        final Uint8List vPlane = image.planes[2].bytes;

        int height = image.height;
        int width = image.width;

        final Uint8List combinedPlanes =
            Uint8List(yPlane.length + uPlane.length + vPlane.length);
        combinedPlanes.setRange(0, yPlane.length, yPlane);
        combinedPlanes.setRange(
            yPlane.length, yPlane.length + uPlane.length, uPlane);
        combinedPlanes.setRange(
            yPlane.length + uPlane.length, combinedPlanes.length, vPlane);

        Mat matImage = Mat.create(rows: height, cols: width, type: MatType.CV_8UC3);
        matImage.data.setAll(0, combinedPlanes);

        Mat bgrBytes = cvtColor(matImage, COLOR_YUV2BGR_NV21);

        Mat blurredImage = gaussianBlur(bgrBytes, (size, size), sigma);

        return Result(processesImage: image, data: {});
    }

`

**Comment by rainyl** [comment]
@TheoNanu Sorry for the late reply, I am kind of busy recently.

I havn't used the camera plugin so not familiar with it, but `Mat.data` actually stores pixel values, ensure `image.planes[0].bytes;` are raw pixel values too.

I will try to provide a camera interaction example in awesome-opencv_dart when having some time, you can also open a PR to https://github.com/rainyl/awesome-opencv_dart if you solved it.

Also, although not recommended, you can use the videoio module from opencv too, which alows opening a `VideoCapture` from a camera, but in this way, you need to use `Texture` if showing the frames is needed.

**Reply by TheoNanu** [reply]
Hello @rainyl, thanks for your reply. Unfortunately, I did not find a solution for this, maybe @gau-nernst has an example, since the initial question mentions pixel values as `Uint8List` which is easy to obtain from a `CameraImage` object. I would prefer to not grab the images using the `VideoCapture` class.

**Reply by rainyl** [reply]
Okay, then I will add an example some time, but can't promise when. You can subscribe https://github.com/rainyl/awesome-opencv_dart/issues/8

**Reply by TheoNanu** [reply]
Ok, I have found a solution for converting the `CameraImage` image object in YUV420 format to `Mat`:

 ```
  int width = image.width;
  int height = image.height;

  // Plane data
  final yPlane = image.planes[0];
  final uPlane = image.planes[1];
  final vPlane = image.planes[2];

  // Calculate stride and copy Y data
  Uint8List yBuffer = Uint8List(width * height);
  for (int row = 0; row < height; row++) {
    yBuffer.setRange(
      row * width,
      (row + 1) * width,
      yPlane.bytes
          .sublist(row * yPlane.bytesPerRow, row * yPlane.bytesPerRow + width),
    );
  }

  // Calculate stride and copy UV data
  int uvWidth = width ~/ 2;
  int uvHeight = height ~/ 2;

  Uint8List uBuffer = Uint8List(uvWidth * uvHeight);
  Uint8List vBuffer = Uint8List(uvWidth * uvHeight);

 // Struture of bytes in plane: uvHeight * (width + padding)
  // bytesPerRow = width + padding
  // bytesPerPixel = distance between two consecutive chroma values
  // For YUV420_888 the first byte represents the chroma value and the second byte represents the padding (only used for memory consistency)

  // Bytes associated with last row do not have the row padding and also the last chroma value does not have padding

  for (int row = 0; row < uvHeight; row++) {
    for (int col = 0; col < uvWidth; col++) {
      int sourceIndex =
          row * uPlane.bytesPerRow + col * (uPlane.bytesPerPixel ?? 1);
      uBuffer[row * uvWidth + col] =
          uPlane.bytes[sourceIndex]; // Use the first byte
      vBuffer[row * uvWidth + col] =
          vPlane.bytes[sourceIndex]; // Use the first byte
    }
  }

 Uint8List yuvBuffer =
      Uint8List(yBuffer.length + uBuffer.length + vBuffer.length);
  yuvBuffer.setRange(0, yBuffer.length, yBuffer);
  yuvBuffer.setRange(yBuffer.length, yBuffer.length + uBuffer.length, uBuffer);
  yuvBuffer.setRange(
      yBuffer.length + uBuffer.length, yuvBuffer.length, vBuffer);

 Mat matImage = Mat.create(
      rows: height + height ~/ 2, cols: width, type: MatType.CV_8UC1);
  matImage.data.setAll(0, yuvBuffer);

  Mat bgrImage = cvtColor(matImage, COLOR_YUV2BGR_IYUV);
```

For some reason the image is rotated to a degree which depends on the current orientation of the device, if you need to correct that you can use `rotate` function from the library with the degree of rotation parameter appropriate to the current orientation

**Reply by rainyl** [reply]
Thanks for your solution~
Actually I have managed to make it work, refer to
https://github.com/rainyl/awesome-opencv_dart/pull/9
But some bugs still exists, so it is not merged.

**Reply by eramirez3** [reply]
Is there any way to make YUV420 to RGB more efficient ?

**Reply by angelru** [reply]
```
static imglib.Image _convertToImage(CameraImage cameraImage) {
    try {
      if (Platform.isAndroid) {
        final int width = cameraImage.width;
        final int height = cameraImage.height;
        final int uvRowStride = cameraImage.planes[1].bytesPerRow;
        final int uvPixelStride = cameraImage.planes[1].bytesPerPixel ?? 1;

        var img = imglib.Image(width: width, height: height);

        for (int y = 0; y < height; y++) {
          final int yRow = y * cameraImage.planes[0].bytesPerRow;

          final int uvRow = (y ~/ 2) * uvRowStride;

          for (int x = 0; x < width; x++) {
            final int yIndex = yRow + x;

            final int uvIndex = uvRow + (x ~/ 2) * uvPixelStride;

            final int yp = cameraImage.planes[0].bytes[yIndex];
            final int up = cameraImage.planes[1].bytes[uvIndex];
            final int vp = cameraImage.planes[2].bytes[uvIndex];

            int r = (yp + 1.402 * (vp - 128)).round().clamp(0, 255);
            int g = (yp - 0.344136 * (up - 128) - 0.714136 * (vp - 128))
                .round()
                .clamp(0, 255);
            int b = (yp + 1.772 * (up - 128)).round().clamp(0, 255);

            img.setPixelRgba(x, y, r, g, b, 255);
          }
        }

        return imglib.copyRotate(img, angle: 90);
      } else {
        return imglib.Image.fromBytes(
          width: cameraImage.width,
          height: cameraImage.height,
          bytes: cameraImage.planes[0].bytes.buffer,
          order: imglib.ChannelOrder.bgra,
          format: imglib.Format.uint8,
        );
      }
    } catch (e) {
      throw Exception('Error en conversión de imagen: $e');
    }
  }
```

Camera:


```
      _cameraController = CameraController(
        camera,
        ResolutionPreset.medium,
        enableAudio: false,
        imageFormatGroup:
            Platform.isAndroid
                ? ImageFormatGroup.yuv420
                : ImageFormatGroup.bgra8888,
      );
```


Usage:
```
var imgC = _convertToImage(cameraImage);
img = await cv.imdecodeAsync(imglib.encodePng(imgC), cv.IMREAD_GRAYSCALE);
```
This always works for me, but I'm not sure it's the "most efficient" way.

Any ideas?

**Reply by angelru** [reply]
The conversion in the example is faster than the one I was using, so I've switched to the one you posted.
